### PR TITLE
Revisit Recursive Relationships for Menu

### PR DESCRIPTION
**Pull request by KatrukOV**
Added validateRecursiveRelationship in MenuItemCustomPersistenceHandler

**Link to QA issue**
[QA-4643](https://github.com/BroadleafCommerce/QA/issues/4643)

### CODE REVIEW

File: src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java [hunk 1]
```diff
@@ -195,7 +191,17 @@ public Entity add(PersistencePackage persistencePackage, DynamicEntityDao dynami
         if (url != null && StringUtils.isEmpty(url.getValue())) {
             url.setValue(" ");
         }
-        return helper.getCompatibleModule(OperationType.BASIC).add(persistencePackage);
+        try {
```
Comment (troian88) [inline src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java:199]: Can we extract this code to some method?

File: src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java [hunk 2]
```diff
@@ -218,8 +234,67 @@ protected FilterMapping createFilterMappingForProperty(String targetPropertyName
...
+    }
+
+    protected Entity validateMenuItem(final Entity entity) throws ServiceException {
+        final Entity entityAfterValidateRecursiveRelationship = this.validateRecursiveRelationship(entity);
+        return this.validateCategory(entityAfterValidateRecursiveRelationship);
     }
+
+    protected Entity validateCategory(Entity entity) {
+        final String currentCategoryId = entity.getPMap().get("id").getValue();
+        final String defaultCategoryId = entity.getPMap().get("linkedMenu").getValue();
```
Comment (troian88) [inline src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java]: It seems we have constant for String literal `"linkedMenu"`

File: src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java [hunk 3]
```diff
@@ -218,8 +234,67 @@ protected FilterMapping createFilterMappingForProperty(String targetPropertyName
...
+        if (currentCategoryId != null && currentCategoryId.equals(defaultCategoryId)) {
+            entity.addValidationError(LINKED_MENU_PROPERTY, "validationCategory");
+        }
+        return entity;
+    }
+
+    protected Entity validateRecursiveRelationship(Entity entity) throws ServiceException {
+        try {
+            String linkedMenuId = entity.findProperty(LINKED_MENU_PROPERTY).getValue();
```
Comment (troian88) [inline src/main/java/org/broadleafcommerce/menu/admin/server/handler/MenuItemCustomPersistenceHandler.java]: I think we can Parse "long" only here and in the following methods use just Long

**Comment by riteshadhikari17** [comment]
Validation looks ok but it is breaking UI. The modal shows all the fields even the hidden ones. Also lets change the message from "Do not add self link" to "Cannot add itself"
image

**Review by riteshadhikari17 — APPROVED** [review]

**Pull request merged**